Code/Framework/AzCore/Tests/DOM/DomPrefixTreeTests.cpp

```diff
@@ -143,62 +231,104 @@ namespace AZ::Dom::Tests
+        EXPECT_EQ(1, VisitorResultCount());
+        EXPECT_TRUE(ValidateResult(Path("/foo"), 99));
+    }
+
+    TEST_F(DomPrefixTreeTests, EarlyTerminateVisit)
+    {
+        size_t visitCalls = 0;
+        auto visitorFn = [&visitCalls](const Path&, int)
         {
-            results.emplace_back(path, n);
+            ++visitCalls;
+            return false;
         };
+        m_runData->m_visitorTree.VisitPath(Path(), visitorFn, PrefixTreeTraversalFlags::TraverseLeastToMostSpecific);
+        EXPECT_EQ(visitCalls, 1);
+        visitCalls = 0;
+        m_runData->m_visitorTree.VisitPath(Path(), visitorFn, PrefixTreeTraversalFlags::TraverseMostToLeastSpecific);
+        EXPECT_EQ(visitCalls, 1);
+    }
 
-        auto validateResult = [&results](const Path& path, int n)
+    TEST_F(DomPrefixTreeTests, RemoveChildrenDuringVisit)
+    {
+        // Remove children during a depth-first traversal
+        auto visitorFn = [&](const Path& path, int)
         {
-            for (const auto& pair : results)
-            {
-                if (pair.first == path)
-                {
-                    return pair.second == n;
-                }
-            }
-            return false;
+            m_runData->m_visitorTree.EraseValue(path, true);
+            return true;
         };
+        m_runData->m_visitorTree.VisitPath(Path(), visitorFn, PrefixTreeTraversalFlags::TraverseMostToLeastSpecific);
 
-        tree.SetValue(Path("/foo"), 99);
-        tree.SetValue(Path("/foo/0"), 0);
-        tree.SetValue(Path("/foo/1"), 42);
-        tree.SetValue(Path("/bar/bat"), 1);
-        tree.SetValue(Path("/bar/baz"), 2);
-
-        tree.VisitPath(Path("/bar"), PrefixTreeMatch::ExactPath, visitorFn);
-        EXPECT_EQ(0, results.size());
-        results.clear();
-
-        tree.VisitPath(Path("/foo/0"), PrefixTreeMatch::ExactPath, visitorFn);
-        EXPECT_EQ(1, results.size());
-        EXPECT_TRUE(validateResult(Path("/foo/0"), 0));
-        results.clear();
-
-        tree.VisitPath(Path("/foo/1"), PrefixTreeMatch::ExactPath, visitorFn);
-        EXPECT_EQ(1, results.size());
-        EXPECT_TRUE(validateResult(Path("/foo/1"), 42));
-        results.clear();
-
-        tree.VisitPath(Path("/foo"), PrefixTreeMatch::SubpathsOnly, visitorFn);
-        EXPECT_EQ(2, results.size());
-        EXPECT_TRUE(validateResult(Path("/foo/0"), 0));
-        EXPECT_TRUE(validateResult(Path("/foo/1"), 42));
-        results.clear();
-
-        tree.VisitPath(Path("/foo"), PrefixTreeMatch::PathAndSubpaths, visitorFn);
-        EXPECT_EQ(3, results.size());
-        EXPECT_TRUE(validateResult(Path("/foo"), 99));
-        EXPECT_TRUE(validateResult(Path("/foo/0"), 0));
-        EXPECT_TRUE(validateResult(Path("/foo/1"), 42));
-        results.clear();
+        // Ensure all entries were successfully deleted
+        TestVisitPath(Path(), PrefixTreeTraversalFlags::None);
+        EXPECT_EQ(0, VisitorResultCount());
```

Comment:
@noortor This is the exact case I was looking in RemoveOrigin.
This make sure that all the children elements are deleted as well.